### DISCUSSION TITLE
v0.20.0

### DISCUSSION THREAD

**Discussion by proditis** [echoCTF/echoCTF.RED#598] [Announcements]
* Significant improvements on cron commands
* Improve frontend performance (reduce the number of queries performed)
* Show failed logins on player profile
* Make backend forms that require player_id into ajax search
* Display better errors when pulling and starting images
* Add precision to stream timestamp
* Reduce CSRF validation errors and turn them into alerts instead of error pages
* Improve error handling on nginx templates
* Player settings update avatar, updates the first time
* Support for private instances
* Improve on demand powerups
* Frontend indicator when connected to vpn
* Make support staff ids into a sysconfig key
* Fix docker-compose errors that got introduced by mixed encoding
* Fix bugs on ratings, writeups, typos in Inquiry models (thnx @HitmanAlharbi)

<hr /><em>This discussion was created from the release <a href='https://github.com/echoCTF/echoCTF.RED/releases/tag/'>v0.20.0</a>.</em>

**Comment by biggerbananas** [comment]
Hi,

Love the idea of this prjoect - looks awesome. I really want to implenet my own version but have been struggling to get it all working (I apologise I'm a hobbyist here and not a proper dev, so am a bit out of my depth) - having spent the past week tinkering with the docker all in one, I've finally managed to get it deployed on a debian box, however I've fallen at the final hurdle and am a bit stuck on what to do next. 

I've got evertyhing installed on a single box at the moment - and can log in to the backend quite happily, but when trying to login to the frontend I get an internal server error. 

Have checked all the config files and everthing seems fine - I'm guessing an issue with the database or accessing it, but the backend is working? 

I'm going to start again and install fresh but would be grateful for some pointers on where to look... 
![error](https://github.com/echoCTF/echoCTF.RED/assets/144009679/146f79dd-bfaa-4d33-b998-8d9b326f6a1b)

**Reply by proditis** [reply]
Hi there, i'm glad you find the project interesting and sorry for the troubles you're having. 

Check the file `frontend/runtime/logs/app.log` from `echoctfred_frontend` for any relevant errors. Unfortunately i cant see the current URL but i suspect its related to the image permissions.

Also are you using the git repo version or the release files? Also which document did you follow for your setup?

**Reply by biggerbananas** [reply]
Hi,

Thanks for the reply - appreciate the help. 

I'm connecting via IP at the moment as I've not pointed a domain at it - I've included a screenshot with the URL. 

I followed the documentation here: 

https://echoctfred.readthedocs.io/INSTALL-LINUX/

Because I've been having issues with composer resolving dependencies I ended up using debian 10 and installing php 7.4.33. This seems to fix all the errors messages and get evertyhing installed. 

As per the docs I pulled the files from here https://github.com/echoCTF/echoCTF.RED.git

Also attached is a shot of the app.log - you can see the post where I am throwing some random login details.


![app log](https://github.com/echoCTF/echoCTF.RED/assets/144009679/fc830ee4-3eb1-46e7-b3aa-287b7d0b398b)
![error](https://github.com/echoCTF/echoCTF.RED/assets/144009679/b1db5b00-73d8-4e13-bf65-a14e973c2a05)

**Comment by biggerbananas** [comment]
![log](https://github.com/echoCTF/echoCTF.RED/assets/144009679/f80727df-657d-4dfe-a9c6-a0dcb6505a2b)

**Reply by proditis** [reply]
The problem seems to be in the `frontend/config/db.php`, comment out or delete the line `'enableParamLogging'=>true,` and give it another try.

**Comment by biggerbananas** [comment]
Thanks,

Have tried that - sadly not fixed yet, but the error is now slightly
different - so looks like that is the right place to be looking.

Might rebuild from scratch with that line removed and try some other
tweaks.

[image: image.png]

Full db.php:

[image: image.png]

Thanks

**Reply by proditis** [reply]
What is the new error you get? Remove the existing app.log file reproduce the error and attach the complete log, without it its really hard to tell what could be the problem.

**Comment by biggerbananas** [comment]
Here you go - sorry didn't mean to take loads of your time. Will continue
to tinker.

**Reply by proditis** [reply]
Thank you for being considerate, don't worry you're not wasting my time. 
By the way no file came on this or your previous post, i suspect github doesn't support these through email.

**Comment by biggerbananas** [comment]
Odd
It is only a small file though so I'll copy and paste it here. Just doing
some research about the hash function:

2023-09-07 19:16:47
[145.224.66.153][-][5369pf58bvivs4hcoaegbu6hdb][error][yii\base\InvalidArgumentException]
yii\base\InvalidArgumentException: Hash is invalid. in
/var/www/echoCTF.RED/frontend/vendor/yiisoft/yii2/base/Security.php:583
Stack trace:
#0 /var/www/echoCTF.RED/frontend/models/Player.php(153):
yii\base\Security->validatePassword()
#1 /var/www/echoCTF.RED/frontend/models/forms/LoginForm.php(65):
app\models\Player->validatePassword()
#2
/var/www/echoCTF.RED/frontend/vendor/yiisoft/yii2/validators/InlineValidator.php(77):
app\models\forms\LoginForm->validatePassword()
#3
/var/www/echoCTF.RED/frontend/vendor/yiisoft/yii2/validators/Validator.php(260):
yii\validators\InlineValidator->validateAttribute()
#4 /var/www/echoCTF.RED/frontend/vendor/yiisoft/yii2/base/Model.php(368):
yii\validators\Validator->validateAttributes()
#5 /var/www/echoCTF.RED/frontend/models/forms/LoginForm.php(90):
yii\base\Model->validate()
#6 /var/www/echoCTF.RED/frontend/controllers/SiteController.php(194):
app\models\forms\LoginForm->login()
#7 [internal function]: app\controllers\SiteController->actionLogin()
#8
/var/www/echoCTF.RED/frontend/vendor/yiisoft/yii2/base/InlineAction.php(57):
call_user_func_array()
#9
/var/www/echoCTF.RED/frontend/vendor/yiisoft/yii2/base/Controller.php(178):
yii\base\InlineAction->runWithParams()
#10 /var/www/echoCTF.RED/frontend/vendor/yiisoft/yii2/base/Module.php(552):
yii\base\Controller->runAction()
#11
/var/www/echoCTF.RED/frontend/vendor/yiisoft/yii2/web/Application.php(103):
yii\base\Module->runAction()
#12
/var/www/echoCTF.RED/frontend/vendor/yiisoft/yii2/base/Application.php(384):
yii\web\Application->handleRequest()
#13 /var/www/echoCTF.RED/frontend/web/index.php(13):
yii\base\Application->run()
#14 {main}
2023-09-07 19:16:47
[145.224.66.153][-][5369pf58bvivs4hcoaegbu6hdb][info][application] $_GET =
[]

$_POST = [
    '_csrf-red' =>
'ZgjBfbKd0GFrgz5DyAtLAxSdfxEjTcNIKgNI4tn6-08wcvcn4MSKDDS3SgGrXzJ6Jc0cXhYeuzxuQi6mraO6CA=='
    'LoginForm' => [
        'username' => 'test'
        'password' => 'supersecretpassword'
        'rememberMe' => '1'
    ]
    'login-button' => ''
]

$_FILES = []

$_COOKIE = [
    'red' => '5369pf58bvivs4hcoaegbu6hdb'
]

$_SESSION = [
    '__flash' => []
    '_csrf-red' => 'Vz6ZRYZm_4tBcTyy1PcO5SxtDAfDtYAG'
]

$_SERVER = [
    'REDIRECT_STATUS' => '200'
    'HTTP_HOST' => '45.76.143.116'
    'HTTP_USER_AGENT' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64;
rv:109.0) Gecko/20100101 Firefox/117.0'
    'HTTP_ACCEPT' =>
'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8'
    'HTTP_ACCEPT_LANGUAGE' => 'en-GB,en;q=0.5'
    'HTTP_ACCEPT_ENCODING' => 'gzip, deflate'
    'CONTENT_TYPE' => 'application/x-www-form-urlencoded'
    'CONTENT_LENGTH' => '245'
    'HTTP_ORIGIN' => 'http://45.76.143.116'
    'HTTP_CONNECTION' => 'keep-alive'
    'HTTP_REFERER' => 'http://45.76.143.116/login'
    'HTTP_COOKIE' => 'red=5369pf58bvivs4hcoaegbu6hdb'
    'HTTP_UPGRADE_INSECURE_REQUESTS' => '1'
    'HTTP_DNT' => '1'
    'HTTP_SEC_GPC' => '1'
    'PATH' => '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'
    'SERVER_SIGNATURE' => '<address>Apache/2.4.38 (Debian) Server at
45.76.143.116 Port 80</address>
'
    'SERVER_SOFTWARE' => 'Apache/2.4.38 (Debian)'
    'SERVER_NAME' => '45.76.143.116'
    'SERVER_ADDR' => '45.76.143.116'
    'SERVER_PORT' => '80'
    'REMOTE_ADDR' => '145.224.66.153'
    'DOCUMENT_ROOT' => '/var/www/echoCTF.RED/frontend/web/'
    'REQUEST_SCHEME' => 'http'
    'CONTEXT_PREFIX' => ''
    'CONTEXT_DOCUMENT_ROOT' => '/var/www/echoCTF.RED/frontend/web/'
    'SERVER_ADMIN' => '[no address given]'
    'SCRIPT_FILENAME' => '/var/www/echoCTF.RED/frontend/web/index.php'
    'REMOTE_PORT' => '25841'
    'REDIRECT_URL' => '/login'
    'GATEWAY_INTERFACE' => 'CGI/1.1'
    'SERVER_PROTOCOL' => 'HTTP/1.1'
    'REQUEST_METHOD' => 'POST'
    'QUERY_STRING' => ''
    'REQUEST_URI' => '/login'
    'SCRIPT_NAME' => '/index.php'
    'PHP_SELF' => '/index.php'
    'REQUEST_TIME_FLOAT' => 1694110607.8003
    'REQUEST_TIME' => 1694110607
]

**Comment by biggerbananas** [comment]
Ok - so I have made some progress.

After commenting out that line in in db.php I was still unable to log my
test user into the front end. Seeing the hash error I thought I would try
creating a user from the command line as per the guide:

./backend/yii player/register username email fullname password offense 1

This appeared to work - but no player had been created when looking in
the admin backend. Interestingly though when trying to log in as that
user instead of a server error it returns invalid password.

So I deleted the original user and created a new one - I can now log
in but get a 403 error as a welcome page:

[image: image.png]

Sorry - no idea how I've managed to make it behave so strangely!.

**Reply by proditis** [reply]
The screenshot is not visible again :smiling_face_with_tear: 

I am not quite sure i understood what happened before in order to try and reproduce it but here are some things that you can try.

Can you provide the following details
* What is the URL you're getting the 403 error? 
* Is the avatar for the player visible when you get the error?
* Do you get the same error if you visit the `http://yourhost/profile/me` or `http://yourhost/help`?

**Reply by biggerbananas** [reply]
No probs visiting those URLs.

The issue arose on login - avatars are visible but it showed a 403 error with a note about not having permissions to perform that action - but I'd not set a homepage in the config. Adding /dashboard has solved that but I'm still seeing server errors. Just created a challenge and now when my test player visits the challenge... 
![error](https://github.com/echoCTF/echoCTF.RED/assets/144009679/3ac6936c-e09e-4522-9e3f-968f0a6859a5)

**Reply by biggerbananas** [reply]
[app.log](https://github.com/echoCTF/echoCTF.RED/files/12552754/app.log)

**Reply by proditis** [reply]
Oh ok, so the 403 was with empty homepage, that i can test to see what is going on :smiley: (opened an issue to investigate that #993)

With regards the "internal server error" 
* clear the app.log
* reload the url to cause the error again
* paste the lines that include the trace output of the error

**Reply by proditis** [reply]
Ok found the issue, the challenge has no questions associated with it (this is a bug that I will have to fix, feel free to track it at #994).

Until this is fixed, visit the backend Gameplay -> Questions and create one for this challenge. Should make it go away

**Reply by biggerbananas** [reply]
I created a challenge and attached a file - then get an error when clicking on the challenge as a user. 

Log attached for that one. 

[app.log](https://github.com/echoCTF/echoCTF.RED/files/12552866/app.log)

I also created challenge without an attached file in case that was causing a problem - same issue, here is the log when visiting challenge #2
[app2.log](https://github.com/echoCTF/echoCTF.RED/files/12552881/app2.log)

**Reply by biggerbananas** [reply]
Ah - great, was about to try that. Sorry should have done that first!

**Reply by proditis** [reply]
Unfortunately, the documentation is seriously lacking, especially for these details, so i dont blame you for not knowing. I see this entire discussion as a huge opportunity to improve the documentation a little bit :smile:

**Comment by biggerbananas** [comment]
And adding /dashboard to the config fixes that issue...

Will have a play and let you know how I get on - really appreciate the
help, looking forward to getting stuck in with it.

**Comment by biggerbananas** [comment]
Happy to try and help with that - I'm trying to log what I'm doing so I can
repeat it. Currently building on a cloud VM - but will probably be
rebuilding on real hardware at some point.

Another hopefully quick one - which I might answer in a sec... I can't get
it to attach a file to the challenge. I've noted it needs to have a name
before it will attach, but then I get file not found when trying to open
the file. I see it is shown as shown in /uploads/filename - I'm thinking it
either isn't uploading as I expect, or there is a permissions issue - just
going to see if I can find where that is on the filesystem now.

**Reply by proditis** [reply]
The attachment and the filename are not exactly working that way (its a bit counter intuitive i admit). Usually the frontend and backend is not on the same host so in that case you just add the filename and upload the file on the frontend by say ssh.

When both frontend and backend are on the same host then the upload folder for both point to the same location so when you upload on backend its becomes available to the frontend also. For now you have to copy the challenge to the frontend/web/uploads folder by hand.

**Comment by biggerbananas** [comment]
Ok - kind of makes sense. When I get to production build I hope to do
separate frontend / backend so good to know.

I don't see an uploads folder though - so I'm going to create it and see
what happens....

**Reply by biggerbananas** [reply]
![uploads](https://github.com/echoCTF/echoCTF.RED/assets/144009679/af1ead42-f939-4545-b03b-1416fe884b5b)

**Reply by proditis** [reply]
Yes you have to create it or you can change it into something else but make sure you also change the `challenge_home` sysconfig key.

**Reply by biggerbananas** [reply]
Thanks - yep have created it and it is working now. 👍

**Reply by biggerbananas** [reply]
I'll leave you in peace for a bit now. I may well be in touch for more help if that's OK though - as I've tried and failed with the VPN server already - so that is next on the list. 

Then docker repo and servers. Excited to get it working though so will keep notes as I go.

**Reply by proditis** [reply]
It was my pleasure, feel free to start a new discussion thread with any issues you have with the VPN.

**Reply by biggerbananas** [reply]
Thanks. Much appreciated.
